File: backend/pkg/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"regexp"
	"sync"
	"time"

	"golang.org/x/crypto/hkdf"
)

// Global Regexp Matchers for URL-Safe Base64
var (
	// Standart Base64'ten URL-Safe'e dönüşüm için
	urlSafeBase64Matcher  = regexp.MustCompile(`\+`) // + karakterini eşler
	urlSafeBase64Matcher2 = regexp.MustCompile(`/`)  // / karakterini eşler
	noPaddingMatcher      = regexp.MustCompile(`=`)  // = karakterini eşler (padding)

	// URL-Safe'ten Standart Base64'e dönüşüm için
	standardMatcher  = regexp.MustCompile(`\-`) // - karakterini eşler
	standardMatcher2 = regexp.MustCompile(`\_`) // _ karakterini eşler
)

// KeyCache yapısı - Thread-safe için sync.RWMutex eklendi
type KeyCache struct {
	sync.RWMutex
	keys    map[string]cachedKey
	maxSize int
}

type cachedKey struct {
	key       []byte
	timestamp time.Time
}

var globalKeyCache = NewKeyCache(1000)

// NewKeyCache yeni bir anahtar önbelleği oluşturur
func NewKeyCache(maxSize int) *KeyCache {
	return &KeyCache{
		keys:    make(map[string]cachedKey),
		maxSize: maxSize,
	}
}

// DeriveKeys JWT token ve session ID kullanarak AES-256 anahtarı türetir
func DeriveKeys(token, sessionId string) ([]byte, error) {
	if token == "" || sessionId == "" {
		return nil, errors.New("anahtar türetme için token ve session ID gerekli")
	}

	cacheKey := fmt.Sprintf("%s|%s", token, sessionId)

	// 1. Kontrol: Read Lock ile hızlıca kontrol
	globalKeyCache.RLock()
	if cached, exists := globalKeyCache.keys[cacheKey]; exists {
		if time.Since(cached.timestamp) < time.Hour { // Anahtar ömrü 1 saat
			globalKeyCache.RUnlock()
			return cached.key, nil
		}
		// Süresi dolmuş, Read Lock'ı serbest bırak
		globalKeyCache.RUnlock()
		// Yazma kilidini al, eski anahtarı silme ve yeni anahtar hesaplama için devam et
	} else {
		globalKeyCache.RUnlock()
	}

	// 2. Kontrol (Double-Check Locking) ve Yazma İşlemi: Write Lock ile senkronizasyon
	globalKeyCache.Lock()
	defer globalKeyCache.Unlock()

	// Yeniden kontrol: Başka bir goroutine bizim Lock'ımızı beklerken anahtarı eklemiş olabilir
	if cached, exists := globalKeyCache.keys[cacheKey]; exists {
		if time.Since(cached.timestamp) < time.Hour {
			// Cache'e yeni eklenmiş, doğrudan dön
			return cached.key, nil
		}
		// Eski kayıt tekrar süresi dolmuşsa silinir
		delete(globalKeyCache.keys, cacheKey)
	}

	// Anahtar türetme işlemi
	masterKey := []byte(token + sessionId)
	hkdfReader := hkdf.New(sha256.New, masterKey, nil, nil)
	derivedKey := make([]byte, 32)
	if _, err := io.ReadFull(hkdfReader, derivedKey); err != nil {
		return nil, fmt.Errorf("HKDF anahtar türetme hatası: %v", err)
	}

	// Cache boyut kontrolü ve temizleme
	if len(globalKeyCache.keys) >= globalKeyCache.maxSize {
		// En eski kaydı bulup sil
		var oldestKey string
		var oldestTime time.Time
		for key, cached := range globalKeyCache.keys {
			if oldestTime.IsZero() || cached.timestamp.Before(oldestTime) {
				oldestTime = cached.timestamp
				oldestKey = key
			}
		}
		if oldestKey != "" {
			delete(globalKeyCache.keys, oldestKey)
		}
	}

	globalKeyCache.keys[cacheKey] = cachedKey{
		key:       derivedKey,
		timestamp: time.Now(),
	}

	return derivedKey, nil
}

// DecryptData base64 şifreli veriyi çözer
func DecryptData(encryptedBase64, token, sessionId string) (map[string]interface{}, error) {
	key, err := DeriveKeys(token, sessionId)
	if err != nil {
		return nil, fmt.Errorf("anahtar türetme hatası: %w", err)
	}

	encryptedData, err := base64.StdEncoding.DecodeString(encryptedBase64)
	if err != nil {
		return nil, errors.New("base64 decode başarısız")
	}

	if len(encryptedData) < 13 {
		return nil, errors.New("şifreli veri çok kısa")
	}

	nonce := encryptedData[:12]
	ciphertext := encryptedData[12:]

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, errors.New("AES şifre oluşturma başarısız")
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, errors.New("GCM modu oluşturma başarısız")
	}

	plaintext, err := aesgcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		// Hata detayını gizle (Oracle Attack Koruması)
		return nil, errors.New("şifre çözme veya doğrulama başarısız")
	}

	var result map[string]interface{}
	if err := json.Unmarshal(plaintext, &result); err != nil {
		return nil, errors.New("JSON parse başarısız")
	}

	// Replay attack koruması - timestamp kontrolü
	if err := validateTimestamp(result); err != nil {
		// Hata detayını gizle (Oracle Attack Koruması)
		return nil, errors.New("timestamp doğrulama başarısız")
	}

	return result, nil
}

// validateTimestamp replay attack koruması için timestamp kontrolü
func validateTimestamp(data map[string]interface{}) error {
	timestampVal, exists := data["_timestamp"]
	if !exists {
		// Detaylı hata verme
		return errors.New("güvenlik damgası (timestamp) eksik")
	}

	timestamp, ok := timestampVal.(float64)
	if !ok {
		return errors.New("geçersiz güvenlik damgası formatı")
	}

	requestTime := time.Unix(0, int64(timestamp)*int64(time.Millisecond))
	now := time.Now()

	// 5 dakikadan eski istekleri reddet (Replay Attack Koruması)
	if now.Sub(requestTime) > 5*time.Minute {
		return errors.New("istek zaman aşımına uğradı")
	}

	// 5 saniyeden yeni istekleri reddet (Clock Skew Koruması)
	if requestTime.Sub(now) > 5*time.Second {
		return errors.New("istemci saati çok ileri")
	}

	return nil
}

// EncryptData veriyi şifreler ve base64 string olarak döndürür
func EncryptData(payload interface{}, token, sessionId string) (string, error) {
	key, err := DeriveKeys(token, sessionId)
	if err != nil {
		return "", fmt.Errorf("anahtar türetme hatası: %w", err)
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("JSON marshal hatası: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("AES şifre oluşturma hatası: %w", err)
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("GCM modu oluşturma hatası: %w", err)
	}

	nonce := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("nonce oluşturma hatası: %w", err)
	}

	// GCM ile şifrele, Tag otomatik olarak eklenir
	encryptedData := aesgcm.Seal(nil, nonce, jsonData, nil)
	finalData := append(nonce, encryptedData...)

	return base64.StdEncoding.EncodeToString(finalData), nil
}

// convertUrlSafeToStandard URL güvenli base64'ü standart base64'e dönüştürür
func convertUrlSafeToStandard(urlSafeBase64 string) string {
	standard := urlSafeBase64
	standard = standardMatcher.ReplaceAllString(standard, "+")
	standard = standardMatcher2.ReplaceAllString(standard, "/")

	// Base64 padding (Eşittir işaretleri) ekle
	switch len(standard) % 4 {
	case 2:
		standard += "=="
	case 3:
		standard += "="
	}
	return standard
}

// DecryptQueryParams şifreli query parametrelerini çözer
func DecryptQueryParams(encryptedQuery, token, sessionId string) (map[string]interface{}, error) {
	standardBase64 := convertUrlSafeToStandard(encryptedQuery)

	// DecryptData artık genel hata döndürdüğü için loglamayı burada yapmayız.
	decryptedParams, err := DecryptData(standardBase64, token, sessionId)
	if err != nil {
		// Hata detayını gizle ve generic bir hata mesajı döndür.
		return nil, errors.New("query parametre çözme/doğrulama başarısız")
	}

	return decryptedParams, nil
}

// EncryptQueryParams query parametrelerini şifreler
func EncryptQueryParams(params map[string]interface{}, token, sessionId string) (string, error) {
	encryptedData, err := EncryptData(params, token, sessionId)
	if err != nil {
		return "", fmt.Errorf("query parametre şifreleme hatası: %w", err)
	}

	// URL güvenli Base64 formatına çevir
	urlSafeBase64 := encryptedData
	urlSafeBase64 = urlSafeBase64Matcher.ReplaceAllString(urlSafeBase64, "-")
	urlSafeBase64 = urlSafeBase64Matcher2.ReplaceAllString(urlSafeBase64, "_")

	// Padding (=) işaretlerini kaldır
	return noPaddingMatcher.ReplaceAllString(urlSafeBase64, ""), nil
}

File: frontend/src/utils/crypto.js
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { randomBytes } from "@noble/hashes/utils";

// Anahtar önbelleği
const keyCache = new Map();

/**
 * JWT Token ve Session ID kullanarak AES-256 anahtarı türetir
 */
async function deriveEncryptionKey(token, sessionId) {
  if (!token || !sessionId) {
    throw new Error(
      "Şifreleme anahtarı türetilemedi: Token veya Session ID eksik"
    );
  }

  const cacheKey = `${token}|${sessionId}`;

  if (keyCache.has(cacheKey)) {
    return keyCache.get(cacheKey);
  }

  try {
    const inputKeyMaterial = new TextEncoder().encode(token + sessionId);
    const derivedKeyBytes = hkdf(
      sha256,
      inputKeyMaterial,
      undefined,
      undefined,
      32
    );

    const result = {
      aesKey: derivedKeyBytes,
      derivedAt: Date.now(),
    };

    // Basit bir önbellek temizleme mekanizması (Go'daki gibi gelişmiş değil ama işe yarar)
    if (keyCache.size > 100) {
      // En eski 50 kaydı sil
      const keysToDelete = Array.from(keyCache.keys()).slice(0, 50);
      keysToDelete.forEach((key) => keyCache.delete(key));
    }

    keyCache.set(cacheKey, result);
    return result;
  } catch (error) {
    console.error("Anahtar türetme hatası:", error);
    throw new Error("Şifreleme anahtarı oluşturulamadı");
  }
}

/**
 * Veriyi AES-GCM ile şifreler
 */
export async function encryptData(data, token, sessionId) {
  try {
    const { aesKey } = await deriveEncryptionKey(token, sessionId);

    // Replay attack koruması için timestamp ve nonce ekle
    const payloadWithTimestamp = {
      ...data,
      _timestamp: Date.now(),
      _nonce: Array.from(randomBytes(8))
        .map((byte) => byte.toString(16).padStart(2, "0"))
        .join(""),
    };

    const dataStr = JSON.stringify(payloadWithTimestamp);
    const dataBytes = new TextEncoder().encode(dataStr);

    const iv = randomBytes(12);

    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      aesKey,
      { name: "AES-GCM" },
      false,
      ["encrypt"]
    );

    const encryptedBuffer = await crypto.subtle.encrypt(
      { name: "AES-GCM", iv: iv },
      cryptoKey,
      dataBytes
    );

    // IV ve şifreli veriyi birleştir
    const combined = new Uint8Array(iv.length + encryptedBuffer.byteLength);
    combined.set(iv);
    combined.set(new Uint8Array(encryptedBuffer), iv.length);

    // Base64 olarak dönüştür
    return btoa(String.fromCharCode(...combined));
  } catch (error) {
    console.error("Veri şifreleme hatası:", error);
    throw new Error("Veri şifrelenemedi: " + error.message);
  }
}

/**
 * Şifreli veriyi çözer
 */
export async function decryptData(encryptedBase64, token, sessionId) {
  try {
    const { aesKey } = await deriveEncryptionKey(token, sessionId);

    const binaryString = atob(encryptedBase64);
    const combined = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
      combined[i] = binaryString.charCodeAt(i);
    }

    if (combined.length < 13) {
      throw new Error("Şifreli veri çok kısa");
    }

    const iv = combined.slice(0, 12);
    const ciphertext = combined.slice(12);

    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      aesKey,
      { name: "AES-GCM" },
      false,
      ["decrypt"]
    );

    const decryptedBuffer = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: iv },
      cryptoKey,
      ciphertext
    );

    const decryptedStr = new TextDecoder().decode(decryptedBuffer);
    return JSON.parse(decryptedStr);
  } catch (error) {
    console.error("Veri çözme hatası:", error);
    // Hata mesajını daha genel tutmak güvenlik açısından daha iyidir
    throw new Error("Yanıt çözme hatası: İstemci şifre çözme başarısız oldu.");
  }
}

/**
 * Query parametrelerini şifreler
 */
export async function encryptQueryParams(params, token, sessionId) {
  try {
    // Query parametrelerine de timestamp ve nonce ekle
    const paramsWithTimestamp = {
      ...params,
      _timestamp: Date.now(),
      _nonce: Array.from(randomBytes(4))
        .map((byte) => byte.toString(16).padStart(2, "0"))
        .join(""),
    };

    const encryptedData = await encryptData(
      paramsWithTimestamp,
      token,
      sessionId
    );

    // URL güvenli Base64 formatına çevir
    const urlSafeBase64 = encryptedData
      .replace(/\+/g, "-")
      .replace(/\//g, "_")
      .replace(/=/g, "");

    return urlSafeBase64;
  } catch (error) {
    console.error("Query parametre şifreleme hatası:", error);
    throw new Error("Query parametreleri şifrelenemedi: " + error.message);
  }
}

/**
 * Şifrelenmiş query parametrelerini çözer
 */
export async function decryptQueryParams(encryptedQuery, token, sessionId) {
  try {
    // URL güvenli Base64'ten standart Base64'e çevir
    const standardBase64 = encryptedQuery
      .replace(/\-/g, "+")
      .replace(/\_/g, "/")
      .padEnd(
        encryptedQuery.length + ((4 - (encryptedQuery.length % 4)) % 4),
        "="
      );

    // Not: Frontend, query yanıtını çözerken timestamp/nonce kontrolü yapmaz.
    // Bu kontrol, Backend'deki Request Body/Query şifre çözme aşamasında yapılır.
    return await decryptData(standardBase64, token, sessionId);
  } catch (error) {
    console.error("Query parametre çözme hatası:", error);
    throw new Error("Query parametreleri çözülemedi: " + error.message);
  }
}

/**
 * Önbelleği temizler
 */
export function clearKeyCache() {
  keyCache.clear();
  console.log("Şifreleme anahtarı önbelleği temizlendi");
}
